extensions/amp-autocomplete/0.1/autocomplete-binding-inline.js

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ownProperty} from '../../../src/utils/object';
+import {tryFocus} from '../../../src/dom';
+
+/**
+ * Inline implementation of autocomplete. This supports autocompleting
+ * multiple input values as part of a larger freeform input element.
+ * @implements {./autocomplete-binding-def.AutocompleteBindingDef}
+ * @private
+ */
+export class AutocompleteBindingInline {
+  /**
+   * Stores the regex match value associated with the portion of the user input to suggest against.
+   * For use when "inline_" is true.
+   */
+  constructor() {
+    /** @private {?RegExpResult} */
+    this.match_ = null;
+  }
+
+  /**
+   * Returns true if a match on the publisher-provided trigger is found in the input element value.
+   * Otherwise, should not display any suggestions.
+   * @param {string} trigger
+   * @param {!HTMLInputElement} inputEl
+   * @return {boolean}
+   */
+  shouldAutocomplete(trigger, inputEl) {
+    const match = this.getClosestPriorMatch_(trigger, inputEl);
+    this.match_ = match;
+    return !!match;
+  }
+
+  /**
+   * Finds the closest string in the user input prior to the cursor
+   * to display suggestions.
+   * @param {string} trigger
+   * @param {!HTMLInputElement} inputEl
+   * @return {?RegExpResult}
+   * @private
+   */
+  getClosestPriorMatch_(trigger, inputEl) {
+    if (trigger === '') {
+      return null;
+    }
+
+    const delimiter = trigger.replace(/([()[{*+.$^\\|?])/g, '\\$1');
```

Comment:
For the constant, you should probably generate it lazily to prevent blocking script initialization:

```js
import {once} from '../../../src/utils/function';

const getThisRegexIdkWhatToNameIt = once(() => /([()[{*+.$^\\|?])/g);

// ...

const delimiter = trigger.replace(getThisRegexIdkWhatToNameIt(), '\\$1');
```

Reply:
...it might also be a good idea to cache the regex created from `delimiter` as well. We're currently compiling each regex roughly on a factor of every keystroke. If the proposed function reads from a top-level delimiter-regex cache map you'll shave off some valuable time from the frame budget.

Reply:
(I'll put on my `wg-performance` hat (🤠) later and do some quick profiling to figure out whether it makes a difference.)

Reply:
Hmm, we don't need to define the regex every time actually. I'll make some changes accordingly, but otherwise what do you think about optimizing the regex in a follow up PR? Since this is under experimental I feel comfortable pushing a basic implementation and making incremental changes.

Reply:
As discussed offline, we can postpone making this decision.